pyroute2/wiset.py

```diff
@@ -0,0 +1,477 @@
+
+>>> wiset = list_ipset("foo")
+>>> wiset.add("1.2.3.4/24")
+
+Listing entries is also easier using WiSet, since it parses for you
+netlink messages:
+
+>>> wiset.content
+{'1.2.3.0/24': IPStats(packets=None, bytes=None, comment=None, timeout=None)}
+'''
+
+import errno
+import uuid
+from collections import namedtuple
+from inspect import getcallargs
+from socket import AF_INET
+
+from pyroute2 import IPSet
+from pyroute2.netlink.exceptions import IPSetError
+from pyroute2.netlink.nfnetlink.ipset import IPSET_FLAG_WITH_COUNTERS
+from pyroute2.netlink.nfnetlink.ipset import IPSET_FLAG_WITH_COMMENT
+
+
+# Debug variable to detect netlink socket leaks
+COUNT = {"count": 0}
+
+
+def need_ipset_socket(fun):
+    """ Decorator to create netlink socket if needed.
+
+    In many of our helpers, we need to open a netlink socket. This can
+    be expensive for someone using many times the functions: instead to have
+    only one socket and use several requests, we will open it again and again.
+
+    This helper allow our functions to be flexible: the caller can pass an
+    optional socket, or do nothing. In this last case, this decorator
+    will open a socket for the caller (and close it after call)
+
+    It also help to mix helpers. One helper can call another one: the socket
+    will be opened only once. We just have to pass the ipset variable.
+
+    Note that all functions using this helper *must* use ipset as variable
+    name for the socket.
+    """
+    def wrap(*args, **kwargs):
+        callargs = getcallargs(fun, *args, **kwargs)
+        if callargs["socket"] is None:
+            # This variable is used only to debug leak in tests
+            COUNT['count'] += 1
+            with IPSet() as socket:
+                callargs["socket"] = socket
+                # We must pop kwargs here, else the function will receive
+                # a dict of dict
+                if "kwargs" in callargs:
+                    callargs.update(callargs.pop("kwargs"))
+                return fun(**callargs)  # pylint:disable=star-args
+
+        return fun(*args, **kwargs)
+
+    return wrap
+
+
+IPStats = namedtuple("IPStats", ["packets", "bytes", "comment", "timeout"])
+
+
+# pylint: disable=too-many-instance-attributes
+class WiSet(object):
+    """ Main Wifirst ipset manipulation class.
+
+    You can do all what you need with this class, you don't need other
+    helpers of this module.
+
+    For example, you can create and an entry in a ipset just with:
+
+    >>> with WiSet(name="mysuperipset") as myset:
+    >>>    myset.create()        # add the ipset in the kernel
+    >>>    myset.add("8.8.8.8")  # add this IP of google to the set
+
+    Netlink sockets are opened by __enter__ and __exit__ function, so you don't
+    have to manage it manually if you use the "with" keyword.
+
+    If you want to manage it manually (for example for long operation in
+    a daemon), you can do the following:
+
+    >>> myset = Wiset(name="mysuperipset")
```

Comment:
`Wiset` or `WiSet`? It may be better to keep it consistent.

Reply:
Indeed, I will fix that